ownCloud/Settings/Passcode/UnlockPasscodeManager.swift

```diff
@@ -0,0 +1,118 @@
+//
+//  UnlockPasscodeManager.swift
+//  ownCloud
+//
+//  Created by Javier Gonzalez on 06/05/2018.
+//  Copyright © 2018 ownCloud GmbH. All rights reserved.
+//
+
+import UIKit
+import ownCloudSDK
+
+let DateHomeButtonPressedKey = "date-home-button-pressed"
+
+class UnlockPasscodeManager: NSObject {
+
+    private var passcodeViewController: PasscodeViewController?
+    private var userDefaults: UserDefaults?
+
+    static var sharedUnlockPasscodeManager : UnlockPasscodeManager = {
+        let sharedInstance = UnlockPasscodeManager()
+        return (sharedInstance)
+    }()
+
+    public override init() {
+        self.userDefaults = UserDefaults.standard
+
+        super.init()
+    }
+
+    // MARK: - Unlock device
+
+    func showPasscodeIfNeeded(viewController: UIViewController, hiddenOverlay:Bool) {
+
+        if isPasscodeActivated() {
+
+            storeDateHomeButtonPressed()
+
+            if self.passcodeViewController == nil {
+                self.passcodeViewController = PasscodeViewController(mode: PasscodeInterfaceMode.unlockPasscode, hiddenOverlay:false, handler: {
+                    DispatchQueue.main.async {
+                        self.userDefaults?.removeObject(forKey: DateHomeButtonPressedKey)
+                        viewController.dismiss(animated: true, completion: nil)
+                        self.passcodeViewController = nil
+                    }
+                })
+
+                viewController.present(self.passcodeViewController!, animated: false, completion: nil)
+            } else {
+                self.passcodeViewController?.showOverlay()
+            }
+        }
+    }
+
+    // MARK: - Interface updates
+
+    private func hideOverlay() {
+        if self.passcodeViewController != nil {
+            self.passcodeViewController?.hideOverlay()
+        }
+    }
+
+    func dismissAskedPasscodeIfDateToAskIsLower() {
+
+        if !isNeccesaryShowPasscode() {
+            if self.passcodeViewController != nil {
+                self.passcodeViewController?.dismiss(animated: true, completion: nil)
+                self.passcodeViewController = nil
+                self.userDefaults?.removeObject(forKey: DateHomeButtonPressedKey)
+            }
+        } else {
+            hideOverlay()
+        }
+    }
+
+    // MARK: - Utils
+
+    private func isPasscodeActivated() -> Bool {
+
+        var output: Bool = true
+
+        if OCAppIdentity.shared().keychain.readDataFromKeychainItem(forAccount: passcodeKeychainAccount, path: passcodeKeychainPath) == nil {
+
+            output = false
+        }
+
+        return output
+    }
+
+    private func isNeccesaryShowPasscode() -> Bool {
```

Comment:
This could also be a read-only property `shouldBeLocked`.

Reply:
Done